peps/pep-0694.rst

```diff
@@ -799,6 +830,50 @@ This means providing the metadata request again to retrieve a new upload resourc
 Clients **MUST NOT** assume that the previous upload resource URL can be reused after deletion.
 
 
+.. _file-upload-completion-callback:
+
+File Upload Completion Callbacks
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Servers **MAY** support asynchronous callbacks to notify clients when file upload processing is complete,
+eliminating the need for polling.
+This capability is advertised through the ``file-upload-completion-callback`` field
+in the :ref:`File Upload Session response <file-upload-session-response>`.
+
+When a server supports callbacks and a client provides a ``callback_url`` in the
+:ref:`completion request <file-upload-session-completion>`,
+the server **SHOULD** POST to the callback URL when processing completes,
+whether successfully or with an error.
+
+The callback request includes the same response body as the
+:ref:`File Upload Session status response <file-upload-session-response>`:
+
+.. code-block:: text
+
+    POST /webhook/upload-complete/12345 HTTP/1.1
+    Host: example.com
+    Content-Type: application/vnd.pypi.upload.v2+json
+
+The body contains the standard File Upload Session status response,
+which includes the ``status`` field indicating ``"complete"`` or ``"error"``,
+along with all other session details.
+
+**Implementation Notes**:
+
+- Callback URLs **MUST** use HTTPS
+- Clients **SHOULD** respond to callbacks quickly (within 5 seconds)
+- Servers **MAY** retry failed callback attempts but **MUST NOT** block
```

Comment:
Is there a way for users to test their callbacks, find out if they're failing, or otherwise debug it?  I'm thinking about similar webhook services like GitHub and GitLab (and ReadTheDocs?) which have a way to trigger a test callback.